lib/jsonWebToken.js

```diff
@@ -0,0 +1,186 @@
+ * @callback
+ */
+exports.verify = function(jwtString, PMEKey, options, callback) {
+
+  /*********************************************************************
+   * Checking parameters
+   ********************************************************************/
+
+  // All of the four parameters are required
+  if (!jwtString)
+    throw new Error('jwtString must be provided in jsonWebToken.verify');
+  if (!PMEKey)
+    throw new Error('PMEKey must be provided in jsonWebToken.verify');
+  if (!options)
+    throw new Error('options must be provided in jsonWebToken.verify');
+  if (!callback)
+    throw new Error('callback must be provided in jsonWebToken.verify');
+  if (typeof callback !== 'function')
+    throw new Error('callback in jsonWebToken.verify must be a function');
+
+  // asynchronous wrapper for callback
+  var done = function() {
+    var args = Array.prototype.slice.call(arguments, 0);
+    return process.nextTick(function() {
+      callback.apply(null, args);
+    });
+  };
+
+  // make sure we have the required fields in options
+  if (!options.audience)
+    return done(new Error('options.audience is missing in jsonWebToken.verify'));
+  if (!options.algorithms)
+    return done(new Error('options.algorithms is missing in jsonWebToken.verify'));
+  if (!Array.isArray(options.algorithms) || options.algorithms.length == 0)
+    return done(new Error('options.algorithms must be an array containing at least 1 algorithm'));
+
+  /*********************************************************************
+   * Checking jwtString structure, getting header, payload and signature
+   ********************************************************************/
+
+  // split jwtString, make sure we have three parts and we have signature
+  var parts = jwtString.split('.');
+  if (parts.length !== 3)
+    return done(new Error('jwtString is malformed'));
+  if (parts[2] === '')
+    return done(new Error('signature is missing in jwtString'));
+  
+  // decode jwsString
+  var decodedToken;
+  try {
+    decodedToken = jws.decode(jwtString);
+  } catch(err) {
+    return done(new Error('invalid token'));
+  }
+
+  if (!decodedToken) {
+    return done(new Error('invalid token'));
+  }
+
+  // get header, payload and signature
+  var header = decodedToken.header;
+  var payload = decodedToken.payload;
+  var signature = decodedToken.signature;
+
+  if (!header)
+    return done(new Error('missing header in the token'));
+  if (!payload)
+    return done(new Error('missing payload in the token'));
+  if (!signature)
+    return done(new Error('missing signature in the token'));
+
+  /*********************************************************************
+   * validate algorithm and signature
+   ********************************************************************/
+
+  // alg in the token header should be one of those provided in options.algorithms
+  if (!~options.algorithms.indexOf(header.alg)) {
```

Comment:
please change this to 
`if (options.algorithms.indexOf(header.alg) === -1) {`
The tilde is tricky, but it's also hard to read.